cplusplus/gen-operators.py

```diff
@@ -1,212 +1,257 @@
+_REQUIRED = 1
+_INPUT = 16
+_OUTPUT = 32
+_DEPRECATED = 64
+_MODIFY = 128
+
+# for VipsOperationFlags
+_OPERATION_DEPRECATED = 8
+
+
+def get_cpp_type(gtype):
+    """Map a gtype to C++ type name we use to represent it.
+    """
+    if gtype in gtype_to_cpp:
+        return gtype_to_cpp[gtype]
+
+    fundamental = gobject_lib.g_type_fundamental(gtype)
+
     # enum params use the C name as their name
-    if GObject.type_is_a(param_enum, prop):
-        return prop.value_type.name
+    if fundamental == GValue.genum_type:
+        return type_name(gtype)
 
-    return gtype_to_cpp[prop.value_type.name]
+    if fundamental in gtype_to_cpp:
+        return gtype_to_cpp[fundamental]
 
-def find_required(op):
-    required = []
-    for prop in op.props:
-        flags = op.get_argument_flags(prop.name)
-        if not flags & Vips.ArgumentFlags.REQUIRED:
-            continue
-        if flags & Vips.ArgumentFlags.DEPRECATED:
-            continue
+    return '<unknown type>'
 
-        required.append(prop)
 
-    def priority_sort(a, b):
-        pa = op.get_argument_priority(a.name)
-        pb = op.get_argument_priority(b.name)
+# swap any '-' for '_'
+def cppize(name):
+    return name.replace('-', '_')
 
-        return pa - pb
 
-    required.sort(priority_sort)
+def generate_operation(operation_name, declaration_only=False):
+    op = Operation.new_from_name(operation_name)
 
-    return required
+    # we are only interested in non-deprecated args
+    args = [[name, flags] for name, flags in op.get_args()
+            if not flags & _DEPRECATED]
 
-# find the first input image ... this will be used as "this"
-def find_first_input_image(op, required):
-    found = False
-    for prop in required:
-        flags = op.get_argument_flags(prop.name)
-        if not flags & Vips.ArgumentFlags.INPUT:
-            continue
-        if GObject.type_is_a(vips_type_image, prop.value_type):
-            found = True
+    # find the first required input image arg, if any ... that will be self
+    member_x = None
+    for name, flags in args:
+        if ((flags & _INPUT) != 0 and
+                (flags & _REQUIRED) != 0 and
+                op.get_typeof(name) == GValue.image_type):
+            member_x = name
             break
 
-    if not found:
-        return None
+    required_input = [name for name, flags in args
+                      if (flags & _INPUT) != 0 and
+                      (flags & _REQUIRED) != 0 and
+                      name != member_x]
+
+    required_output = [name for name, flags in args
+                       if ((flags & _OUTPUT) != 0 and
+                           (flags & _REQUIRED) != 0) or
+                       ((flags & _INPUT) != 0 and
+                        (flags & _REQUIRED) != 0 and
+                        (flags & _MODIFY) != 0) and
+                       name != member_x]
+
+    has_output = len(required_output) >= 1
+
+    # TODO: Should we output the operation description as comment?
```

Comment:
Would this be useful?

Reply:
Sure, why not. I'm sure some C++ users would like it.

Reply:
Fixed with: https://github.com/libvips/libvips/pull/1220/commits/241f52fb54fa32ecc1f36238badcd255d69d5283.

The parameters are also described with that commit.